blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/MetricIndexData.java

```diff
@@ -0,0 +1,204 @@
+package com.rackspacecloud.blueflood.io;
+
+import java.util.*;
+
+/**
+ * When we ingest these below metrics, ES generates the following indexes
+ *
+ *     metric           indexes
+ *   -----------        -------
+ *   foo.bar.baz.aux -> [foo, foo.bar, foo.bar.baz, foo.bar.baz.aux, bar, baz, aux]
+ *   foo.bar.baz     -> [foo, foo.bar, foo.bar.baz, bar, baz]
+ *
+ * If we request ES to return us upto next two levels of the metric indexes for foo.bar,
+ * we get this below response from ES where key is the metric index and doc_count is
+ * the number of documents(think metric names) it points to.
+ *
+ * "buckets" : [ {
+ *     "key" : "foo.bar",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz.aux",
+ *     "doc_count" : 1
+ * } ]
+ *
+ * If we feed each metric index to this class by calling the add(metricIndex, docCount) method,
+ * this class analyzes the data and provides methods to retrieve different aspects of this data
+ * with respect to baseLevel(if set to 2 with above data, uses foo.bar as reference).
+ *
+ */
+public class MetricIndexData {
```

Comment:
this class encapsulates the functionality a lot better.  I'm on the cusp of understanding the algorithm!